src/sentry/discover/compare_tables.py

```diff
@@ -0,0 +1,193 @@
+from sentry.models.organization import Organization
+from sentry.models.project import Project
+from sentry.search.eap.types import EAPResponse, SearchResolverConfig
+from sentry.search.events.types import SAMPLING_MODES, EventsResponse, SnubaParams
+from sentry.snuba import spans_rpc
+from sentry.snuba.metrics_performance import query as metrics_query
+
+logger = logging.getLogger(__name__)
+
+
+class CompareTableResult(Enum):
+    BOTH_FAILED = "both_requests_failed"
+    FIELD_NOT_FOUND = "field_not_found"
+    METRICS_FAILED = "metrics_failed"
+    NO_FIELDS = "no_fields"
+    NO_PROJECT = "no_project"
+    PASSED = "passed"
+    SNQL_EAP_FAILED = "snql_eap_failed"
+
+
+class CompareTableResultDict(TypedDict):
+    passed: bool
+    reason: CompareTableResult
+    fields: list[str] | None
+    widget_query: DashboardWidgetQuery
+    mismatches: list[str] | None
+
+
+def compare_table_results(metrics_query_result: EventsResponse, snql_eap_result: EAPResponse):
+    eap_data_row = snql_eap_result["data"][0] if len(snql_eap_result["data"]) > 0 else {}
+    metrics_data_row = (
+        metrics_query_result["data"][0] if len(metrics_query_result["data"]) > 0 else {}
+    )
+    mismatches = []
+    for field, data in metrics_data_row.items():
+        if is_equation(field):
+            continue
+        translated_field, *rest = translate_columns([field])
+        if data is not None and eap_data_row[translated_field] is None:
+            logger.info("Field %s not found in EAP response", field)
+            sentry_sdk.set_tag("dashboard_comparison_mismatch_field", field)
+            mismatches.append(field)
+
+    return (len(mismatches) == 0, mismatches)
+
+
+def compare_tables_for_dashboard_widget_queries(
+    widget_query: DashboardWidgetQuery,
+) -> CompareTableResultDict:
+    widget: DashboardWidget = widget_query.widget
+    dashboard: Dashboard = widget.dashboard
+    organization: Organization = dashboard.organization
+    projects: list[Project] = list(dashboard.projects.all())
+    if len(list(projects)) == 0:
+        return {
+            "passed": False,
+            "reason": CompareTableResult.NO_PROJECT,
+            "fields": None,
+            "widget_query": widget_query,
+            "mismatches": None,
+        }
+
+    project = projects[0]
+
+    fields = widget_query.fields
+    if len(fields) == 0:
+        return {
+            "passed": False,
+            "reason": CompareTableResult.NO_FIELDS,
+            "fields": None,
+            "widget_query": widget_query,
+            "mismatches": None,
+        }
+
+    snuba_dataclass = _get_snuba_dataclass_for_dashboard_widget(widget, list(projects))
+    selected_columns = _get_field_list(fields)
+    equations = [equation for equation in _get_equation_list(widget_query.fields or []) if equation]
+    query = widget_query.conditions
+
+    has_metrics_error = False
+    has_snql_eap_error = False
+
+    try:
+        metrics_query_result = metrics_query(
```

Comment:
I don't think we can just make a metrics query here - has to be a metrics_enhanced_performance query. Dashboards fallback to indexed transactions when no metrics data is present.